src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/Json/MessageTypeInfoResolver.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            var propertyInfo = CreatePropertyInfo(typeInfo, field.JsonName, field, isSerializable: true);
+            typeInfo.Properties.Add(propertyInfo);
+
+            // We have a property for reading and writing the JSON name so remove from mappings.
+            mappings.Remove(field.JsonName);
+        }
+
+        // Fields have two mappings: the original field name and the camelcased JSON name.
+        // The JSON name can also be customized in proto with json_name option.
+        // Remaining mappings are for extra setter only properties.
+        foreach (var mapping in mappings)
+        {
+            var propertyInfo = CreatePropertyInfo(typeInfo, mapping.Key, mapping.Value, isSerializable: false);
+            typeInfo.Properties.Add(propertyInfo);
+        }
+
+        return typeInfo;
+    }
+
+    private static bool IsStandardMessage(Type type, [NotNullWhen(true)] out MessageDescriptor? messageDescriptor)
+    {
+        if (!typeof(IMessage).IsAssignableFrom(type))
+        {
+            messageDescriptor = null;
+            return false;
+        }
+
+        messageDescriptor = JsonConverterHelper.GetMessageDescriptor(type);
+        if (messageDescriptor == null)
+        {
+            return false;
+        }
+
+        // Wrappers and well known types are handled by converters.
+        if (ServiceDescriptorHelpers.IsWrapperType(messageDescriptor))
+        {
+            return false;
+        }
+        if (JsonConverterHelper.WellKnownTypeNames.ContainsKey(messageDescriptor.FullName))
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private JsonPropertyInfo CreatePropertyInfo(JsonTypeInfo typeInfo, string name, FieldDescriptor field, bool isSerializable)
+    {
+        var propertyInfo = typeInfo.CreateJsonPropertyInfo(
+            JsonConverterHelper.GetFieldType(field),
+            name);
+
+        // Properties that don't have this flag set are only used to deserialize incoming JSON.
+        if (isSerializable)
+        {
+            propertyInfo.ShouldSerialize = (o, v) =>
+            {
+                return JsonConverterHelper.ShouldFormatFieldValue((IMessage)o, field, v, !_context.Settings.IgnoreDefaultValues);
+            };
+            propertyInfo.Get = (o) =>
+            {
+                return field.Accessor.GetValue((IMessage)o);
+            };
+        }
+
+        propertyInfo.Set = GetSetMethod(field);
+
+        return propertyInfo;
+    }
+
+    private static Action<object, object?> GetSetMethod(FieldDescriptor field)
+    {
+        if (field.IsMap)
+        {
+            return (o, v) =>
+            {
+                // The serializer creates a collection. Copy contents to collection on read-only property.
+                // An extra collection is being created here that's then thrown away.
+                // This will be removed once S.T.J supports deserializing onto a read-only property.
```

Comment:
Is this referring to https://github.com/dotnet/runtime/issues/30258?

```suggestion
                // This will be removed once S.T.J supports deserializing onto a read-only property.
                // https://github.com/dotnet/runtime/issues/30258
```